### ISSUE TITLE
Question split_mmio_region change start address

### ISSUE THREAD

**Issue by PhilippTakacs** [unicorn-engine/unicorn#1810]
Hi

I'm currently working on some changes reading MemoryRegion handling and found this commend: https://github.com/unicorn-engine/unicorn/blob/7b8c63dfe650b5d4d2bf684526161971925e6350/uc.c#L1225-L1226

I have a solution for this using Aliases and avoiding memcpy. This fix could also used for `split_mmio_region`, but it would change the API, because then the offset in the callback would be relativ to the initial mapped address. What do you think about such a change?

**Comment by wtdcode** [comment]
> > In that case, if the mapping was changed, the offset would be garbage?
> 
> No, the offset would be as if there was no split (by unmap or mem_protect). The change can be best seen in `test_mem_protect_mmio` there is following line:
> 
> https://github.com/unicorn-engine/unicorn/blob/7b8c63dfe650b5d4d2bf684526161971925e6350/tests/unit/test_mem.c#L236
> 
> 
> There the address would change to 0x1020.
> I would argue this is more intuitive. I ask because it would be an API change.
> 
> If you are interested in the current state of the change it's on my fork: https://github.com/PhilippTakacs/unicorn/tree/split

My bad, I designed this offset and wrote the sample before and it took me 5 minutes to understand the comments just now... I got your idea now. Does your branch only affect MMIO regions offsets?

**Comment by PhilippTakacs** [comment]
> Does your branch only affect MMIO regions offsets?

Yes, only the API for for the MMIO read/write callbacks would be changed. Currently not implemented but can be tested by changing the `split_mmio_region` calls to `split_region` on my branch.

**Comment by wtdcode** [comment]
> > Does your branch only affect MMIO regions offsets?
> 
> Yes, only the API for for the MMIO read/write callbacks would be changed. Currently not implemented but can be tested by changing the `split_mmio_region` calls to `split_region` on my branch.

If we are going to break the offset here, I'm considering making it consistent with other memory hooks by passing the actual address instead of the offset. What's your opinion?

Some sidenotes: My original thought was that it could be handy because users don't have to save the base addresses and they only care about the offsets in most cases for relative usages (like memory registers etc).

**Comment by PhilippTakacs** [comment]
> > > Does your branch only affect MMIO regions offsets?
> > 
> > 
> > Yes, only the API for for the MMIO read/write callbacks would be changed. Currently not implemented but can be tested by changing the `split_mmio_region` calls to `split_region` on my branch.
> 
> If we are going to break the offset here, I'm considering making it consistent with other memory hooks by passing the actual address instead of the offset. What's your opinion?
> 
> Some sidenotes: My original thought was that it could be handy because users don't have to save the base addresses and they only care about the offsets in most cases for relevant usages (like memory registers etc).

I would keep the offset with this explanation. To have the base address in the callbacks the user_data can be used.

Sidenode: I don't use/need MMIO. I have asked this question to have my changes clean and (from my perspective) ready to merge at the time I make the PR
